### PR TITLE
Add adaptive-step ODE solver and VJP calculation with JAX ops

### PR DESCRIPTION
**Pull request by cellweaver**
Dormand-Prince adaptive stepsize ODE integration using JAX control flow operators and full jit compilation. Based on code in #1068.

### CODE REVIEW

**Comment by shoyer** [comment]
Could you keep this as `jax.experimental.ode` for now, so it's clearer what you've changed?

**Comment by dsweaver** [comment]
Done

File: jax/experimental/ode.py
Comment (mattjj) [inline jax/experimental/ode.py]: Let's not accidentally delete this license header!

File: jax/experimental/ode.py
Comment (mattjj) [inline jax/experimental/ode.py]: We should probably move these tests into the `tests/` directory like all the others, but that can wait until after we merge this.

File: jax/experimental/ode.py
Comment (mattjj) [inline jax/experimental/ode.py]: Could we keep this example+plotting code in the file somewhere? That way it'll be easy to factor out into an example file.

**Review by mattjj — APPROVED** [review]
Thanks for this! Using control flow operators is a big step forward. I left a couple comments to take a look at, but otherwise LGTM.

Do you have any timing comparisons before/after that would be interesting to include in this PR comment thread for posterity?

**Comment by dsweaver** [comment]
Thanks for comments! I have recovered the license header and plotting code and I have also added a benchmarking function. My results for `pend_benchmark_odeint()` on CPU (Xeon 2.2GHz) are:

scipy odeint elapsed time (1 of 10): 1.0394439697265625
scipy odeint elapsed time (2 of 10): 1.030179738998413
scipy odeint elapsed time (3 of 10): 1.0403587818145752
scipy odeint elapsed time (4 of 10): 1.009368896484375
scipy odeint elapsed time (5 of 10): 0.9918606281280518
scipy odeint elapsed time (6 of 10): 0.9111509323120117
scipy odeint elapsed time (7 of 10): 0.9812769889831543
scipy odeint elapsed time (8 of 10): 1.0298435688018799
scipy odeint elapsed time (9 of 10): 1.01572847366333
scipy odeint elapsed time (10 of 10): 0.968522310256958
JAX odeint elapsed time (1 of 10): 1.2705836296081543
JAX odeint elapsed time (2 of 10): 0.0016872882843017578
JAX odeint elapsed time (3 of 10): 0.0012602806091308594
JAX odeint elapsed time (4 of 10): 0.0011641979217529297
JAX odeint elapsed time (5 of 10): 0.0011451244354248047
JAX odeint elapsed time (6 of 10): 0.0010805130004882812
JAX odeint elapsed time (7 of 10): 0.0011022090911865234
JAX odeint elapsed time (8 of 10): 0.0010726451873779297
JAX odeint elapsed time (9 of 10): 0.000997781753540039
JAX odeint elapsed time (10 of 10): 0.0011608600616455078
norm(scipy result-jax result): 3.201899927344282e-06

**Comment by mattjj** [comment]
Awesome, thanks!

File: jax/experimental/ode.py
Comment (shoyer) [inline jax/experimental/ode.py]: This looks a little strange to me. The usual reason for using `lax.cond` is to avoid computing the outputs in each branch (hence why `cond` requires the `true_fun`/`false_fun` arguments) but you aren't doing that here.

If I recall, `cond` also has some issues with supporting higher level transformation like `vmap`.

So I think it would probably be better to use the simpler `np.where` here (and everyone else below).

File: jax/experimental/ode.py
Comment (shoyer) [inline jax/experimental/ode.py]: It is better to use `np.concatenate` once (on a Python list of arrays) rather than to use `np.append` inside a loop, which has to reallocate the full result each time.

**Comment by dsweaver** [comment]
Done and thanks for the comments. Changing `runge_kutta_step()` to JAX control flow improved compilation time significantly.

scipy odeint elapsed time (1 of 10): 0.827517032623291
scipy odeint elapsed time (2 of 10): 0.8226215839385986
scipy odeint elapsed time (3 of 10): 0.800774097442627
scipy odeint elapsed time (4 of 10): 0.8159639835357666
scipy odeint elapsed time (5 of 10): 0.9570231437683105
scipy odeint elapsed time (6 of 10): 1.1280930042266846
scipy odeint elapsed time (7 of 10): 1.0953774452209473
scipy odeint elapsed time (8 of 10): 1.1448588371276855
scipy odeint elapsed time (9 of 10): 0.9961385726928711
scipy odeint elapsed time (10 of 10): 0.8309123516082764
JAX odeint elapsed time (1 of 10): 0.5770602226257324
JAX odeint elapsed time (2 of 10): 0.0011138916015625
JAX odeint elapsed time (3 of 10): 0.0009877681732177734
JAX odeint elapsed time (4 of 10): 0.0009739398956298828
JAX odeint elapsed time (5 of 10): 0.001004934310913086
JAX odeint elapsed time (6 of 10): 0.0009517669677734375
JAX odeint elapsed time (7 of 10): 0.0009701251983642578
JAX odeint elapsed time (8 of 10): 0.0009329319000244141
JAX odeint elapsed time (9 of 10): 0.0009241104125976562
JAX odeint elapsed time (10 of 10): 0.0010046958923339844
norm(scipy result-jax result): 3.2018999052375912e-06

File: jax/experimental/ode.py
Comment (shoyer) [reply jax/experimental/ode.py]: also: it would probably be cleaner to test against scipy's solvers.

**Review by shoyer — APPROVED** [review]

**Comment by mattjj** [comment]
Thanks for this contribution! It seems to have improved performance significantly. Let's keep iterating on follow-ups (like factoring out the tests, adding a scipy interface, etc.) Exciting!

**Pull request merged**
